src/util/rangeutil.hpp

```diff
@@ -320,6 +319,41 @@ std::pair<Value, Value> minmax_value(const Seq& seq, Compare cmp = Compare{}) {
 template <typename Map>
 auto keys(Map& m) DEDUCED_RETURN_TYPE(util::transform_view(m, util::first));
 
+// Test if sequence is sorted after apply projection `proj` to elements.
+// (TODO: this will perform unnecessary copies if `proj` returns a reference;
+// specialize on this if it becomes an issue.)
+
+template <
+    typename Seq,
+    typename Proj,
+    typename Compare = std::less<typename std::result_of<Proj (typename sequence_traits<Seq>::value_type)>::type>
+>
+bool is_sorted_by(const Seq& seq, const Proj& proj, Compare cmp = Compare{}) {
+    auto i = std::begin(seq);
+    auto e = std::end(seq);
+
+    if (i==e) return true;
+
+    // Special one-element case for forward iterators.
+    if (is_forward_iterator<decltype(i)>::value) {
+        auto j = i;
+        if (++j==e) return true;
+    }
+
```

Comment:
It would simplify the loop, but I think the more complex current implementation is still comprehensible, and does mean that we can relax the requirements on the projected values and save some operations.

As it's a library routine, I think it is better to err on the side of efficiency and genericity, provided these are exercised in testing.

Regarding the braces on all the `return`s: I'll put them in, but I think it will make the code harder to read, not easier.

Reply:
Oh! Should add though that there is already the implicit requirement that the projected values are move assignable.